test-data/unit/check-literal.test

```diff
@@ -0,0 +1,556 @@
+reveal_type(b)  # E: Revealed type is 'Literal[False]'
+[builtins fixtures/bool.pyi]
+[out]
+
+[case testLiteralBasicBoolUsageTypeAlias]
+from typing_extensions import Literal
+
+at = Literal[True]
+bt = Literal[False]
+a: at
+b: bt
+
+reveal_type(a)  # E: Revealed type is 'Literal[True]'
+reveal_type(b)  # E: Revealed type is 'Literal[False]'
+[builtins fixtures/bool.pyi]
+[out]
+
+[case testLiteralBasicStrUsage]
+from typing_extensions import Literal
+
+a: Literal[""]
+b: Literal["  foo bar  "]
+c: Literal['  foo bar  ']
+
+reveal_type(a)  # E: Revealed type is 'Literal['']'
+reveal_type(b)  # E: Revealed type is 'Literal['  foo bar  ']'
+reveal_type(c)  # E: Revealed type is 'Literal['  foo bar  ']'
+[out]
+
+[case testLiteralBasicStrUsageSlashes]
+from typing_extensions import Literal
+
+a: Literal[r"foo\nbar"]
+b: Literal["foo\nbar"]
+
+reveal_type(a)
+reveal_type(b)
+[out skip-path-normalization]
+main:6: error: Revealed type is 'Literal['foo\\nbar']'
+main:7: error: Revealed type is 'Literal['foo\nbar']'
+
+[case testLiteralBasicNoneUsage]
+from typing_extensions import Literal
+a: Literal[None]
+reveal_type(a)   # E: Revealed type is 'None'
+# Note: Literal[None] and None are equivalent
+[out]
+
+[case testLiteralDisallowAny]
+from typing import Any
+from typing_extensions import Literal
+from missing_module import BadAlias     # E: Cannot find module named 'missing_module' \
+                                        # N: (Perhaps setting MYPYPATH or using the "--ignore-missing-imports" flag would help)
+
+a: Literal[Any]                         # E: Parameter 1 of Literal[...] is of type Any
+b: Literal[BadAlias]                    # E: Parameter 1 of Literal[...] is of type Any
+[out]
+
+[case testLiteralDisallowFloats]
+from typing_extensions import Literal
+a: Literal[3.14]  # E: invalid type comment or annotation
+b: 3.14           # E: invalid type comment or annotation
+[out]
+
+[case testLiteralDisallowFloatsTypeAlias]
+from typing_extensions import Literal
+at = Literal[3.14]  # E: Invalid type alias \
+                    # E: The type "Type[Literal]" is not generic and not indexable
+bt = 3.14
+
+a: at               # E: Invalid type "__main__.at"
+b: bt               # E: Invalid type "__main__.bt"
+[out]
+
+[case testLiteralDisallowComplexNumbers]
+from typing_extensions import Literal
+a: Literal[3j]       # E: invalid type comment or annotation
+b: Literal[3j + 2]   # E: invalid type comment or annotation
+c: 3j                # E: invalid type comment or annotation
+d: 3j + 2            # E: invalid type comment or annotation
+[out]
+
+[case testLiteralDisallowComplexNumbersTypeAlias]
+from typing_extensions import Literal
+at = Literal[3j]        # E: Invalid type alias \
+                        # E: The type "Type[Literal]" is not generic and not indexable
+a: at                   # E: Invalid type "__main__.at"
+[builtins fixtures/complex.pyi]
+[out]
+
+[case testLiteralDisallowComplexExpressions]
```

Comment:
A somewhat meta comment: Since test runtimes are already somewhat painful, I often prefer merging multiple related test cases into a single one, especially if they are pretty straightforward. There is significant per-test case overhead. In the future it may make sense to support explicitly subdividing tests into multiple sections to make it easier to maintain longer test cases.

Reply:
I merged a few of the existing tests together, though don't think I was able to for this specific one (some uses of complex numbers raise an error at the parsing stage while others raise an error at the semantic analysis stage -- errors at the parsing stage appear suppress any subsequent one).

In any case, I'll be sure to keep this in mind as I add new tests + tune the existing ones.